import java.util.HashSet;
import java.util.Scanner;

public class LongestSubstringWithoutRepeatingCharacters {

    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);

        System.out.print("Enter string: ");
        String s = sc.nextLine();

        HashSet<Character> set = new HashSet<>();

        int left = 0;
        int maxLength = 0;

        for (int right = 0; right < s.length(); right++) {

            char current = s.charAt(right);

            while (set.contains(current)) {

                set.remove(s.charAt(left));

                left++;
            }

            set.add(current);

            int length = right - left + 1;

            maxLength = Math.max(maxLength, length);
        }

        System.out.println(
                "Longest substring length = " + maxLength);

        sc.close();
    }
}